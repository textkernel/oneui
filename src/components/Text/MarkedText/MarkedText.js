import * as React from 'react';
import PropTypes from 'prop-types';
import { bem } from '../../../utils';
import { Text } from '../Text';
import styles from './MarkedText.scss';

const { block, elem } = bem('MarkedText', styles);

export const MarkedText = (props) => {
    const { marker, children, ...rest } = props;

    let result = children;
    if (marker) {
        const escapedMarker = marker.replace(/[-[\]{}()*+?.,^$|#]/g, '\\$&');
        const re = new RegExp(`(${escapedMarker})`, 'gi');
        result = children.split(re).map((part, i) =>
            part.toLowerCase() === marker.toLowerCase() ? (
                // eslint-disable-next-line react/no-array-index-key
                <mark key={`match${i}`} {...elem('marked', props)} role="contentinfo">
                    {part}
                </mark>
            ) : (
                part
            )
        );
    }

    return (
        <Text {...rest} {...block(props)}>
            {result}
        </Text>
    );
};

MarkedText.displayName = 'MarkedText';

MarkedText.propTypes = {
    /** Text content */
    children: PropTypes.string.isRequired,
    /** The text that should be marked if found within the children */
    marker: PropTypes.string.isRequired,
};
