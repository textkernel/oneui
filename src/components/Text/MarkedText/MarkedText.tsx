import * as React from 'react';
import { bem } from '../../../utils';
import { Text, Props as TextProps } from '../Text';
import styles from './MarkedText.scss';

const { block, elem } = bem('MarkedText', styles);

interface Props extends Omit<TextProps, 'children'> {
    /** String that will be rendered as text */
    children: string;
    /** Part of the text that will be bolded */
    marker: string;
}

export const MarkedText = (props: Props) => {
    const { marker, children, ...rest } = props;

    let result: string | (string | React.JSX.Element)[] = children;

    if (marker) {
        const escapedMarker = marker.replace(/[-[\]{}()*+?.,^$|#]/g, '\\$&');
        const re = new RegExp(`(${escapedMarker})`, 'gi');
        result = children.split(re).map((part, i) =>
            part.toLowerCase() === marker.toLowerCase() ? (
                // eslint-disable-next-line react/no-array-index-key
                <mark key={`match${i}`} {...elem('marked', props)} role="contentinfo">
                    {part}
                </mark>
            ) : (
                part
            )
        );
    }

    return (
        <Text {...rest} {...block(props)}>
            {result}
        </Text>
    );
};

MarkedText.displayName = 'MarkedText';
