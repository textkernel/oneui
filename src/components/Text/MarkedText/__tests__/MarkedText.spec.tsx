import React from 'react';
import { render, RenderResult, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { MarkedText } from '../MarkedText';

describe('<MarkedText> that renders a text block while marking matched substrings', () => {
    let view: RenderResult;

    it('should render correctly when there are no matches', () => {
        view = render(<MarkedText marker="else">Some text content</MarkedText>);

        expect(view.container).toMatchSnapshot();
        expect(screen.queryByRole('contentinfo')).not.toBeInTheDocument();
    });

    it('should render correctly when empty marker passed', () => {
        view = render(<MarkedText marker="">Some text content</MarkedText>);

        expect(screen.queryByRole('contentinfo')).not.toBeInTheDocument();
    });

    it('should render correctly when there are matches', () => {
        view = render(<MarkedText marker="text">Some text content</MarkedText>);

        expect(view.container).toMatchSnapshot();
        expect(screen.getByRole('contentinfo')).toBeVisible();
    });

    it('should render correctly with multiple matches', () => {
        const marker = 'te';
        view = render(
            <MarkedText marker={marker}>Some text content that has text more then once</MarkedText>
        );

        expect(view.container).toMatchSnapshot();
        expect(screen.getAllByRole('contentinfo')).toHaveLength(3);
        expect(screen.getAllByRole('contentinfo')[0].innerHTML).toBe(marker);
    });

    it('should mark special characters', () => {
        const marker = ".?!#$%^&()[]:;\\?''";
        view = render(<MarkedText marker={marker}>{marker}</MarkedText>);

        expect(view.container).toMatchSnapshot();
        expect(screen.getByRole('contentinfo')).toBeVisible();
    });

    it('should pass props to Text', () => {
        view = render(
            <MarkedText marker="text" inline context="brand">
                Some text content
            </MarkedText>
        );

        expect(view.container).toMatchSnapshot();
        expect(screen.getByRole('contentinfo')).toBeVisible();
    });
});
