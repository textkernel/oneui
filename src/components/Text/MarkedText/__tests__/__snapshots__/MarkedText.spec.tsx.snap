// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MarkedText> that renders a text block while marking matched substrings should mark special characters 1`] = `
<div>
  <p>
    <mark
      class="MarkedText__marked"
      role="contentinfo"
    >
      .?!#$%^&()[]:;\\?''
    </mark>
  </p>
</div>
`;

exports[`<MarkedText> that renders a text block while marking matched substrings should pass props to Text 1`] = `
<div>
  <span
    class="Text--context_primary"
  >
    Some 
    <mark
      class="MarkedText__marked"
      role="contentinfo"
    >
      text
    </mark>
     content
  </span>
</div>
`;

exports[`<MarkedText> that renders a text block while marking matched substrings should render correctly when there are matches 1`] = `
<div>
  <p>
    Some 
    <mark
      class="MarkedText__marked"
      role="contentinfo"
    >
      text
    </mark>
     content
  </p>
</div>
`;

exports[`<MarkedText> that renders a text block while marking matched substrings should render correctly when there are no matches 1`] = `
<div>
  <p>
    Some text content
  </p>
</div>
`;

exports[`<MarkedText> that renders a text block while marking matched substrings should render correctly with multiple matches 1`] = `
<div>
  <p>
    Some 
    <mark
      class="MarkedText__marked"
      role="contentinfo"
    >
      te
    </mark>
    xt con
    <mark
      class="MarkedText__marked"
      role="contentinfo"
    >
      te
    </mark>
    nt that has 
    <mark
      class="MarkedText__marked"
      role="contentinfo"
    >
      te
    </mark>
    xt more then once
  </p>
</div>
`;
